### PR TITLE
Jormungandr: add impacted stops in /disruptions

### PR DESCRIPTION
**Pull request by antoine-de**
Enhance the disruption object with data on the impacted stoptimes

TODO: for the moment no information on the base schedule is outputed

``` json
{
    "disruptions": 
    [
        {
            "status": "past",
            "application_periods": 
            [
                {
                    "begin": "20151201T060400",
                    "end": "20151201T082259"
                }
            ],
            "updated_at": "20151202T090248",
            "uri": "3eabf0a6-2bd6-472f-94c4-3aedd83caaa5",
            "contributor": "ire sncf",
            "severity": 
            {
                "color": "#000000",
                "priority": 42,
                "name": "trip delayed",
                "effect": "DETOUR"
            },
            "cause": "a problem",
            "impacted_objects": [
                {
                    "pt_object": {
                        "vehicle_journey": {
                            "uri": "bob"
                            ...
                        }
                    },
                    "impacted_stops": [
                        {
                            "stop_point": { "uri": "B", ... },
                            "base_departure": 171500,
                            "base_arrival": 170000,
                            "cause": "cochon sur les voies",
                            "stoptime_effect": "DELETED"
                        },
                        {
                            "stop_point": { "uri": "E", ... },
                            "amended_departure": 171500,
                            "amended_arrival": 170000,
                            "cause": "cochon sur les voies",
                            "stoptime_effect": "ADDED"
                        },
                        {
                            "stop_point": { "uri": "C", ... },
                            "amended_departure": 181700,
                            "amended_arrival": 181700,
                            "base_departure": 180000,
                            "base_arrival": 180000,
                            "cause": "vache sur les voies",
                            "stoptime_effect": "SIGNIFICANT_DELAYS"
                        },
                        {
                            "stop_point": { "uri": "D", ... },
                            "amended_departure": 192300,
                            "amended_arrival": 192300,
                            "base_departure": 190000,
                            "base_arrival": 190000,
                            "cause": "vache sur les voies",
                            "stoptime_effect": "SIGNIFICANT_DELAYS"
                        }
                    ]
                }
            ]
        },
        {
            "status": "past",
            "application_periods": 
            [
                {
                    "begin": "20151201T060400",
                    "end": "20151201T082259"
                }
            ],
            "updated_at": "20151202T090248",
            "uri": "3eabf0a6-2bd6-472f-94c4-3aedd83caaa6",
            "contributor": "transilien",
            "severity": 
            {
                "color": "#000000",
                "priority": 42,
                "name": "trip delayed",
                "effect": "NO_SERVICE"
            },
            "cause": "another problem",
            "impacted_objects": [
                {
                    "pt_object": {
                        "line": {
                            "uri": "bobette",
                            ...
                        }
                    }
                },
                {
                    "pt_object": {
                        "stop_area": {
                            "uri": "bobitto",
                            ...
                        }
                    }
                }
            ],
            "messages": [
                {
                    "text": "<p>_pas_de_filtre<br></p>",
                    "channel": 
                    {
                        "content_type": "text/html",
                        "types": ["web"],
                        "name": "we"
                    }
                }
            ]
        }
    ]
}
```

### CODE REVIEW

File: source/jormungandr/tests/check_utils.py [hunk 1]
```diff
@@ -922,10 +922,28 @@ def is_valid_disruption(disruption):
...
+        if get_not_null(pt_obj, 'embedded_type') == 'trip' and effect != 'NO_SERVICE':
             impacted_stops = get_not_null(impacted_obj, 'impacted_stops')
 
-            #TODO
+            assert len(impacted_stops) > 0
+            for impacted_stop in impacted_stops:
+                is_valid_stop_point(get_not_null(impacted_stop, 'stop_point'), depth_check=0)
+
+                get_not_null(impacted_stop, "cause")
+                get_not_null(impacted_stop, "stoptime_effect") in ('ADDED', 'DELETED', 'DELAYED')
```
Comment (xlqian) [inline source/jormungandr/tests/check_utils.py]: assert ?

File: source/type/pb_converter.cpp [hunk 2]
```diff
@@ -123,34 +123,66 @@ static void fill_property(const std::string& name,
...
+            }
+        }
+    }
 };
 
 void fill_pb_object(const nt::disruption::PtObj& ptobj,
+                    const nt::disruption::Impact& impact,
                     const nt::Data& data,
-                    pbnavitia::PtObject* pb_pt_pbj, int max_depth,
+                    pbnavitia::Impact* pb_impact, int max_depth,
```
Comment (xlqian) [inline source/type/pb_converter.cpp:183]: can this pointer be replaced by a ref?

File: source/type/pb_converter.cpp [hunk 3]
```diff
@@ -123,34 +123,66 @@ static void fill_property(const std::string& name,
...
     }
     void operator()(const nt::disruption::UnknownPtObj&) const {}
+    void operator()(const nt::MetaVehicleJourney* mvj) const {
+        auto* pobj = add_pt_object(mvj);
+
+        //for meta vj we also need to output the impacted stoptimes
+        for (const auto& stu: impact.aux_info.stop_times) {
+            auto* impacted_stop = pobj->add_impacted_stops();
+            impacted_stop->set_cause(stu.cause);
+            impacted_stop->set_effect(pbnavitia::DELAYED);
```
Comment (xlqian) [inline source/type/pb_converter.cpp:163]: why delayed?

File: source/type/pb_converter.cpp (hunk 2 above)
Comment (antoine-de) [inline source/type/pb_converter.cpp:183]: the protobuf always return ptr for mutable objects and ref pour const ones.

File: source/type/pb_converter.cpp (hunk 3 above)
Comment (antoine-de) [inline source/type/pb_converter.cpp:163]: for the moment it's the only thing handled

File: source/jormungandr/tests/check_utils.py (hunk 1 above)
Comment (antoine-de) [inline source/jormungandr/tests/check_utils.py]: you're right :wink:

File: source/jormungandr/jormungandr/interfaces/v1/fields.py [hunk 4]
```diff
@@ -757,7 +757,13 @@ def output(self, key, obj):
...
 
 impacted_stop = {
-
+    "stop_point": NonNullNested(stop_point),
+    "base_arrival_time": SplitDateTime(date=None, time='base_stop_time.arrival_time'),
+    "base_departure_time": SplitDateTime(date=None, time='base_stop_time.departure_time'),
+    "amended_arrival_time": SplitDateTime(date=None, time='amended_stop_time.arrival_time'),
+    "amended_departure_time": SplitDateTime(date=None, time='amended_stop_time.departure_time'),
+    "cause": fields.String(),
+    "stop_timeeffect": enum_type()
```
Comment (pbougue) [inline source/jormungandr/jormungandr/interfaces/v1/fields.py]: `stop_time_effect` ?

File: source/jormungandr/tests/check_utils.py [hunk 5]
```diff
@@ -922,10 +922,28 @@ def is_valid_disruption(disruption):
...
+        if get_not_null(pt_obj, 'embedded_type') == 'trip' and effect != 'NO_SERVICE':
             impacted_stops = get_not_null(impacted_obj, 'impacted_stops')
 
-            #TODO
+            assert len(impacted_stops) > 0
+            for impacted_stop in impacted_stops:
+                is_valid_stop_point(get_not_null(impacted_stop, 'stop_point'), depth_check=0)
+
+                get_not_null(impacted_stop, "cause")
+                assert(get_not_null(impacted_stop, "stoptime_effect") in ('ADDED', 'DELETED', 'DELAYED'))
```
Comment (pbougue) [inline source/jormungandr/tests/check_utils.py]: `stop_time_effect` ?

File: source/kraken/apply_disruption.cpp [hunk 6]
```diff
@@ -177,7 +177,10 @@ struct add_impacts_visitor : public apply_impacts_visitor {
                 vj->physical_mode = pt_data.physical_modes[0];
                 vj->physical_mode->vehicle_journey_list.push_back(vj);
                 vj->name = new_vj_uri;
-
+            }
+            // we need to associate the stoptimes to the created vj
+            for (auto& stu: impact->aux_info.stop_times) {
+                stu.stop_time.vehicle_journey = vj;
```
Comment (TeXitoi) [inline source/kraken/apply_disruption.cpp:183]: I don't like that, but I have no proposition

**Comment by TeXitoi** [comment]
OK for me.

File: source/kraken/apply_disruption.cpp (hunk 6 above)
Comment (xlqian) [inline source/kraken/apply_disruption.cpp:183]: maybe a function of impact?

Comment (antoine-de) [inline source/kraken/apply_disruption.cpp:183]: not sure it's really worth it

**Comment by xlqian** [comment]
ok for me

File: source/type/pb_converter.cpp [hunk 7]
```diff
@@ -123,34 +123,67 @@ static void fill_property(const std::string& name,
...
+                               0, now, action_period, show_codes, DumpMessage::No);
+            }
+        }
+    }
 };
 
-void fill_pb_object(const nt::disruption::PtObj& ptobj,
+
+static void fill_pb_object(const nt::disruption::PtObj& ptobj,
+                    const nt::disruption::Impact& impact,
```
Comment (pbougue) [inline source/type/pb_converter.cpp:181]: minor: indent

File: source/type/pb_converter.cpp [hunk 8]
```diff
@@ -511,8 +542,8 @@ void fill_pb_object(nt::Line const* l, const nt::Data& data,
 }
 
 void fill_pb_object(const nt::LineGroup* lg, const nt::Data& data,
-        pbnavitia::LineGroup* line_group, int max_depth,
-        const pt::ptime& now, const pt::time_period& action_period
+                    pbnavitia::LineGroup* line_group, int max_depth,
+                    const pt::ptime& now, const pt::time_period& action_period
```
Comment (pbougue) [inline source/type/pb_converter.cpp:546]: minor: last `,` at end of line

File: source/type/pb_converter.cpp [hunk 9]
```diff
@@ -781,41 +812,41 @@ void fill_pb_object(const nt::VehicleJourney* vj,
 }
 
 
-void fill_pb_object(const nt::StopTime* st, const type::Data &data,
+void fill_pb_object(const nt::StopTime& st, const type::Data &data,
```
Comment (pbougue) [inline source/type/pb_converter.cpp:815]: :+1:

**Comment by pbougue** [comment]
Ok for me (2 minors to report in next PR maybe)

**Pull request merged**